march_rqt_gait_generator/src/march_rqt_gait_generator/gait_generator_controller.py

```diff
@@ -0,0 +1,310 @@
+            self.time_slider_thread.stop()
+            self.time_slider_thread = None
+
+    def update_gait_duration(self, duration):
+        rescale_setpoints = self.view.scale_setpoints_check_box.isChecked()
+
+        if self.subgait.has_setpoints_after_duration(duration) and not rescale_setpoints:
+            if not self.subgait.has_multiple_setpoints_before_duration(duration):
+                self.view.message(title='Could not update gait duration',
+                                  msg='Not all joints have multiple setpoints before duration ' + str(duration))
+                self.view.duration_spin_box.setValue(self.subgait.duration)
+                return
+            discard_setpoints = self.view.yes_no_question(title='Gait duration lower than highest time setpoint',
+                                                          msg='Do you want to discard any setpoints higher than the '
+                                                          'given duration?')
+            if not discard_setpoints:
+                self.view.duration_spin_box.setValue(self.subgait.duration)
+                return
+        self.subgait.set_duration(duration, rescale_setpoints)
+        self.view.time_slider.setRange(0, 100 * self.subgait.duration)
+
+        was_playing = self.time_slider_thread is not None
+        self.stop_time_slider_thread()
+
+        self.view.update_joint_widgets(self.subgait.joints)
+
+        if was_playing:
+            self.start_time_slider_thread()
+
+    def import_gait(self):
+        file_name, f = self.view.open_file_dialogue()
+
+        gait = ModifiableSubgait.from_file(self.robot, file_name, self)
+        if gait is None:
+            rospy.logwarn('Could not load gait %s', file_name)
+            return
+        self.subgait = gait
+        self.view.load_gait_into_ui(self.subgait)
+        for joint in self.subgait.joints:
+            self.connect_plot(joint)
+        self.current_time = 0
+
+        self.gait_directory = '/'.join(file_name.split('/')[:-3])
+        rospy.loginfo('Setting gait directory to %s', str(self.gait_directory))
+        self.view.change_gait_directory_button.setText(self.gait_directory)
+
+        # Clear undo and redo history
+        self.joint_changed_history = RingBuffer(capacity=100, dtype=list)
+        self.joint_changed_redo_list = RingBuffer(capacity=100, dtype=list)
+
+    def export_gait(self):
+        if self.view.mirror_check_box.isChecked():
+            key_1 = self.view.mirror_key1_line_edit.text()
+            key_2 = self.view.mirror_key2_line_edit.text()
+            mirror = self.subgait.get_mirror(key_1, key_2)
+
+            if mirror:
+                self.export_to_file(mirror, self.get_gait_directory())
+            else:
+                self.view.notify('Could not mirror gait', 'Check the logs for more information.')
+                return
+
+        self.export_to_file(self.subgait, self.get_gait_directory())
```

Comment:
I would make this method return the subgait string which should be exported to a file. This makes it easier to test, without having to read files.